### DISCUSSION TITLE
Bad login attempt

### DISCUSSION THREAD

**Discussion by Marcus101RR** [Pryaxis/TShock#2064] [Q&A] [answered]
What is the cause of this issue and how do I repair it? It only seems to affect a select few accounts, but not too many, but a significant amount of people are reporting this issue on my server and I am unable to resolve it. The only solution is for them to start a new account.

**Comment by hakusaro** [comment]
SSC?

**Reply by Marcus101RR** [reply]
Yes

User attempts all options:
/login Name password (correct password) -> Bad login attempt
/login -> Bad login Attempt
/login password -> Bad login attempt
/login wrongpassword -> Incorrect password

configs:
  "DisableLoginBeforeJoin": false,
  "DisableUUIDLogin": false,
  "KickEmptyUUID": true,
  "AllowRegisterAnyUsername": false,
  "AllowLoginAnyUsername": true,

**Reply by hakusaro** [reply]
Where are they seeing the message?

**Comment by hakusaro** [comment]
~It looks like your users aren't in a group that has `tshock.account.login` and therefore can't login. Are they maybe being assigned a group that doesn't have that permission?~

**Reply by Marcus101RR** [reply]
by default they are assigned default and have login, I did create a custom code that migrates them to a custom temp group but inherits (or at least should) their parent group with still having the login node. I suppose I can isolate this bug and see if forcing to add that node will prevent this issue.

EDIT: But this doesn't happen until after login, so wonder why.

**Comment by hakusaro** [comment]
Okay, I'm dumb. Let me try again. It looks like you have `AllowLoginAnyUsername` enabled, and they need to have passed both username and password to login. E.g., `/login maruchan whydoIalwaysrememberwhatcircularisinjapaneseitsbecauseofyoumaru`.

The error only happens if the username is null or empty, and if the username parameter is nil, and the setting is enabled.

![Screen Shot 0002-06-20 at 11 33 34](https://user-images.githubusercontent.com/532647/85209126-3b960400-b2ea-11ea-8be9-cb3fb06dc300.png)

See https://github.com/Pryaxis/TShock/blob/8bd6d72cf75efbef581e5509bdc7927925416952/TShockAPI/Commands.cs#L792. TShock can't continue at this point because it can't locate an account to login to.

**Reply by Marcus101RR** [reply]
Yes but they are providing a Account name and a password and still get bad login attempt, including correct password, but when they provide the incorrect password it intentionally it says incorrect. So clearly tshock DID find the account but for some reason is being a dick about it, why else would it say incorrect when its ACTUALLY incorrect?

VIDEO: https://imgur.com/a/QSCT7je

**Reply by hakusaro** [reply]
The code is written less than great. TShock will attempt to find an account by player name first. So it's located the account. If you pass 2 parameters to the function, you get the error if the first is null. However, if you only pass one parameter, the system will try to log you into the account with the player name's account.

I don't see any way that a user could provide both a username and a password and TShock successfully locating the account that a user passes in (ergo, it's just defaulting to the account at the player name, not the one typed in).

**Reply by hakusaro** [reply]
I did a really poor job of explaining that.

`account` is set to `the player name`
then `account` is redefined to `what someone passed in` if you have the config option set and there are 2 parameters
but if there's not two parameters or the config option is set,
tshock goes with the first thing `account` was set to.

**Reply by Marcus101RR** [reply]
So is the TShock code the bug or is the user making a mistake? Because they provided 100% the Username and password incorrect and correct, and clearly tshock "knowns" its an incorrect password when the user intentionally provides it (as seen in the video). But no matter the case:

My account works fine, if I logout, and do the following:
/login -> Successfully logged in (uuid method)
/login Marcus101RR mypassword -> successfully logged in
/login mypassword -> successfully logged in
/login notmypassword -> incorrect password.

Its inconstant with some users though. So should I be reporting this as a bug, or is there more to it?

**Reply by hakusaro** [reply]
@Marcus101RR are you using a different keyboard layout than en-US? Maybe some kind of fullwidth nonsense is happening? I'm not sure how in the very last attempts it could trip that flow unless the input is being mangled.

Edit: I mean the last couple in the video, where the user goes `/login somename password` => bad login attempt.

**Reply by Marcus101RR** [reply]
Heh keyboard layout, where do I check this. I am in USA, so it should be standard.


EDIT: US (QWERTY)

**Reply by hakusaro** [reply]
The user who has the issue should know what their keyboard layout is set to. E.g., in settings? Is the user typing two spaces after every parameter?

**Reply by Marcus101RR** [reply]
I can show you a video of the same attempt being done on an account this doesn't happen to it nothing fails. So just trying to isolate the inconstancy of why this is happening. It is the main reason why my server is always 5 players online because no one can login.

1. The user is me in the video btw, but it happens to others.
2. I am adding 1 space bar to it. /login SPACE username SPACE password, never failed me.

**Reply by Marcus101RR** [reply]
Video: https://imgur.com/a/k2mMnlo

My account, with every attempt not giving bad attempt.

**Reply by Marcus101RR** [reply]
Analysis: Some users claim they have more than 1 account on my server, would this be the conflict (my theory is) the code is searching for the matching uuid of the account, first one it finds, and they happen to want to login to the second account. But this shouldn't trigger if they provide both username and password.

So I have 3 accounts:
Marcus101RR, Saiyan, Eggman
Can login fine with Marcus101RR or Eggman,
Saiyan however is always bad login attempt.

All with the same uuid, and password. So it seems Saiyan account is corrupted somehow.

**Reply by hakusaro** [reply]
Are you able to confirm if they're playing with vanilla terraria, or a mod loader of some kind?

**Reply by Marcus101RR** [reply]
It is Vanilla (this is all me doing this testing now) with the client and with a server. I have another inconsistency: https://imgur.com/a/fddzXuq

This time I removed Saiyan's TimeActive and UUID to null, and this happens. Eggman still works fine. I think this needs a one on one call because there is a big issue with tshock NOT finding accounts im specifying, or I have to send the sqlite db.

**Reply by hakusaro** [reply]
I think you should separate the two issues: if you're experiencing problems after changing the database UUID field to `null` that's kind of expected, to be honest.

If you can't replicate the issue at all even on the bugged accounts with the right & wrong passwords, I would say the problem can be narrowed down to the client, or a plugin or mod interfering with the chat text before it hits the TShock command processor.

The number one thing that we need to know is whether or not you can reproduce the exact issue without changing anything (ergo, do you _ever_ get the bad login attempt doing the exact same commands?).

The TShock login system attempts to find accounts pretty rigorously. It would also be helpful if you can confirm that the database field for Saiyan's name is just ASCII Saiyan or if there are other characters stored (diacritics, non breaking spaces, etc).

**Reply by Marcus101RR** [reply]
There are other users who can replicate the issue without any interference. It isn't because I modified the db or not, it clearly is an issue that just appears out of the blue for users with alternative accounts. How do I check or send you the information regarding that entry?

**Comment by bartico6** [comment, marked as answer]
Issue was figured out to be a faulty tShock fork by yours truly, commaster

Can be closed.
